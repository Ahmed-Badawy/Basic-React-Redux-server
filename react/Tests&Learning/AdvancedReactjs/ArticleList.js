import React, { PureComponent } from 'react';
import Article from "./Article";

export default class ArticleList extends PureComponent {
  componentWillUpdate = (nextProps, nextState) => {
    console.log(`... ${this.constructor.name} Component is updating... `);
  }  
  
  render() {
    return (
      <div>
        {Object.values(this.props.articles).map(article=>
            <Article key={article.id} article={article} store={this.props.store} />
        )}
      </div>
    )
  }
}
