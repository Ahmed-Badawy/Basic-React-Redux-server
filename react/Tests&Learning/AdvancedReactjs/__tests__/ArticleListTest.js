import React from 'react';
import renderer from "react-test-renderer";

import ArticleList from "../ArticleList";


const testProps = {
    articles: {
        a: { id: "a", title:"a" },
        b: { id: "b", title:"b" }
    },
    store: {
        lookupAuthors: jest.fn( ()=>({}) )
    }
}

// react-test-tenderer : snapshot technique to test the property.
describe("ArticleList", ()=>{
    test("ArticleList Renders Correctly", ()=>{
        const tree = renderer.create(<ArticleList {...testProps} />).toJSON();
        expect(tree).toMatchSnapshot();
        expect(tree.children.length).toBe(2);
    })
})



// Shallow Testing With Enzyme
import {shallow, configure} from "enzyme"; 
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

describe("ArticleList Shallow Testing", ()=>{
    test("ArticleList Renders Correctly with Shallow Snapshots", ()=>{
        const wrapper = shallow(<ArticleList {...testProps} />);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.getElement().props.children.length).toBe(2);
    })
})




