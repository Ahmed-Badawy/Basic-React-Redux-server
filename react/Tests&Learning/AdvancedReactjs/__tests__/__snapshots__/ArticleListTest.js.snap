// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleList ArticleList Renders Correctly 1`] = `
<div>
  <div
    style={
      Object {
        "color": "#444",
        "margin": "14px",
        "padding": "8px",
      }
    }
  >
    <div
      style={
        Object {
          "border": "2px gray solid",
          "borderRadius": "10px",
          "color": "#333",
          "fontSize": "1.5em",
          "paddingLeft": "8px",
        }
      }
    >
      a
    </div>
    <div>
      Invalid Date
    </div>
    <div>
      <a>
         
      </a>
    </div>
    <div />
  </div>
  <div
    style={
      Object {
        "color": "#444",
        "margin": "14px",
        "padding": "8px",
      }
    }
  >
    <div
      style={
        Object {
          "border": "2px gray solid",
          "borderRadius": "10px",
          "color": "#333",
          "fontSize": "1.5em",
          "paddingLeft": "8px",
        }
      }
    >
      b
    </div>
    <div>
      Invalid Date
    </div>
    <div>
      <a>
         
      </a>
    </div>
    <div />
  </div>
</div>
`;

exports[`ArticleList Shallow Testing ArticleList Renders Correctly with Shallow Snapshots 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ArticleList
    articles={
      Object {
        "a": Object {
          "id": "a",
          "title": "a",
        },
        "b": Object {
          "id": "b",
          "title": "b",
        },
      }
    }
    store={
      Object {
        "lookupAuthors": [MockFunction] {
          "calls": Array [
            Array [
              undefined,
            ],
            Array [
              undefined,
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": Object {},
            },
            Object {
              "isThrow": false,
              "value": Object {},
            },
          ],
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Article
          article={
            Object {
              "id": "a",
              "title": "a",
            }
          }
          store={
            Object {
              "lookupAuthors": [MockFunction] {
                "calls": Array [
                  Array [
                    undefined,
                  ],
                  Array [
                    undefined,
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": Object {},
                  },
                  Object {
                    "isThrow": false,
                    "value": Object {},
                  },
                ],
              },
            }
          }
        />,
        <Article
          article={
            Object {
              "id": "b",
              "title": "b",
            }
          }
          store={
            Object {
              "lookupAuthors": [MockFunction] {
                "calls": Array [
                  Array [
                    undefined,
                  ],
                  Array [
                    undefined,
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": Object {},
                  },
                  Object {
                    "isThrow": false,
                    "value": Object {},
                  },
                ],
              },
            }
          }
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "a",
        "nodeType": "class",
        "props": Object {
          "article": Object {
            "id": "a",
            "title": "a",
          },
          "store": Object {
            "lookupAuthors": [MockFunction] {
              "calls": Array [
                Array [
                  undefined,
                ],
                Array [
                  undefined,
                ],
              ],
              "results": Array [
                Object {
                  "isThrow": false,
                  "value": Object {},
                },
                Object {
                  "isThrow": false,
                  "value": Object {},
                },
              ],
            },
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "b",
        "nodeType": "class",
        "props": Object {
          "article": Object {
            "id": "b",
            "title": "b",
          },
          "store": Object {
            "lookupAuthors": [MockFunction] {
              "calls": Array [
                Array [
                  undefined,
                ],
                Array [
                  undefined,
                ],
              ],
              "results": Array [
                Object {
                  "isThrow": false,
                  "value": Object {},
                },
                Object {
                  "isThrow": false,
                  "value": Object {},
                },
              ],
            },
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Article
            article={
              Object {
                "id": "a",
                "title": "a",
              }
            }
            store={
              Object {
                "lookupAuthors": [MockFunction] {
                  "calls": Array [
                    Array [
                      undefined,
                    ],
                    Array [
                      undefined,
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": Object {},
                    },
                    Object {
                      "isThrow": false,
                      "value": Object {},
                    },
                  ],
                },
              }
            }
          />,
          <Article
            article={
              Object {
                "id": "b",
                "title": "b",
              }
            }
            store={
              Object {
                "lookupAuthors": [MockFunction] {
                  "calls": Array [
                    Array [
                      undefined,
                    ],
                    Array [
                      undefined,
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": Object {},
                    },
                    Object {
                      "isThrow": false,
                      "value": Object {},
                    },
                  ],
                },
              }
            }
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "a",
          "nodeType": "class",
          "props": Object {
            "article": Object {
              "id": "a",
              "title": "a",
            },
            "store": Object {
              "lookupAuthors": [MockFunction] {
                "calls": Array [
                  Array [
                    undefined,
                  ],
                  Array [
                    undefined,
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": Object {},
                  },
                  Object {
                    "isThrow": false,
                    "value": Object {},
                  },
                ],
              },
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "b",
          "nodeType": "class",
          "props": Object {
            "article": Object {
              "id": "b",
              "title": "b",
            },
            "store": Object {
              "lookupAuthors": [MockFunction] {
                "calls": Array [
                  Array [
                    undefined,
                  ],
                  Array [
                    undefined,
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": Object {},
                  },
                  Object {
                    "isThrow": false,
                    "value": Object {},
                  },
                ],
              },
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
